reference-implementation/lib/writable-stream.js

```diff
@@ -4,225 +4,661 @@ const { InvokeOrNoop, PromiseInvokeOrNoop, PromiseInvokeOrFallbackOrNoop, Valida
+  assert(stream._writeRequests.length > 0);
+
+  const writeRequest = stream._writeRequests.shift();
+  writeRequest._resolve(undefined);
+}
+
+function WritableStreamUpdateBackpressure(stream, backpressure) {
+  const state = stream._state;
+  const writer = stream._writer;
+
+  if (state === 'writable') {
+    if (backpressure === false) {
+      return;
     }
-    if (this._state === 'closed') {
-      return Promise.reject(new TypeError('cannot close an already-closed stream'));
+
+    stream._state = 'waiting';
+
+    if (writer !== undefined) {
+      WritableStreamDefaultWriterReadyPromiseReset(writer);
     }
-    if (this._state === 'errored') {
-      return Promise.reject(this._storedError);
+
+    return;
+  }
+
+  assert(state === 'waiting');
+
+  if (backpressure === true) {
+    return;
+  }
+
+  stream._state = 'writable';
+
+  if (writer !== undefined) {
+    WritableStreamDefaultWriterReadyPromiseResolve(writer, undefined);
+  }
+}
+
+class WritableStreamDefaultWriter {
+  constructor(stream) {
+    if (IsWritableStream(stream) === false) {
+      throw new TypeError('WritableStreamDefaultWriter can only be constructed with a WritableStream instance');
     }
-    if (this._state === 'waiting') {
-      this._readyPromise_resolve(undefined);
+    if (IsWritableStreamLocked(stream) === true) {
+      throw new TypeError('This stream has already been locked for exclusive writing by another writer');
     }
 
-    this._state = 'closing';
-    EnqueueValueWithSize(this._queue, 'close', 0);
-    CallOrScheduleWritableStreamAdvanceQueue(this);
+    this._ownerWritableStream = stream;
+    stream._writer = this;
+
+    const state = stream._state;
+
+    if (state === 'writable' || state === 'waiting' || state === 'closing') {
+      WritableStreamDefaultWriterClosedPromiseInitialize(this);
+    } else {
+      if (state === 'closed') {
+        WritableStreamDefaultWriterClosedPromiseInitializeAsResolved(this, undefined);
+      } else {
+        assert(state === 'errored', 'state must be errored');
+
+        WritableStreamDefaultWriterClosedPromiseInitializeAsRejected(this, stream._storedError);
+      }
+    }
+
+    if (state === 'waiting') {
+      WritableStreamDefaultWriterReadyPromiseInitialize(this);
+    } else {
+      WritableStreamDefaultWriterReadyPromiseInitializeAsResolved(this, undefined);
+    }
+  }
+
+  get closed() {
+    if (IsWritableStreamDefaultWriter(this) === false) {
+      return Promise.reject(CreateWritableStreamDefaultWriterBrandCheckException('closed'));
+    }
 
     return this._closedPromise;
   }
 
+  get desiredSize() {
+    if (IsWritableStreamDefaultWriter(this) === false) {
+      throw CreateWritableStreamDefaultWriterBrandCheckException('desiredSize');
+    }
+
+    if (this._ownerWritableStream === undefined) {
+      throw CreateWritableStreamDefaultWriterLockException('desiredSize');
+    }
+
+    return WritableStreamDefaultWriterGetDesiredSize(this)
+  }
+
   get ready() {
```

Comment:
It would technically be a normative change as we would stop returning the same promise every time and start returning a new one for each function invocation. But I think that would be OK.